ludwig/datasets/loaders/av_mnist.py

```diff
@@ -0,0 +1,214 @@
+                    elif self.modal == "image":
+                        self.data = self.data.reshape(self.data.shape[0], 1, 28, 28)
+                    self.write_output_dataset(self.labels, self.data, os.path.join(self.raw_dataset_dir, dataset))
+                else:
+                    raise ValueError("the value of modal should be given")
+
+        return super().transform_files(file_paths)
+
+    def load_unprocessed_dataframe(self, file_paths: List[str]) -> pd.DataFrame:
+        """Load dataset files into a dataframe."""
+        return self.output_training_and_test_data()
+
+    def process_source_dataset(self, raw_dataset_dir="."):
+        """An intermediate method to write .npy files for the training and the test image and label data to the raw dataset directory
+        :args:
+            raw_dataset_dir (str) : the directory where all the raw data lives
+        :returns:
+            None"""
+        file_names = {
+            "train_data": "train-images-idx3-ubyte.gz",
+            "train_labels": "train-labels-idx1-ubyte.gz",
+            "test_data": "t10k-images-idx3-ubyte.gz",
+            "test_labels": "t10k-labels-idx1-ubyte.gz",
+        }
+        print("Raw dataset working directory: %s" % raw_dataset_dir)
+
+        for key, file_name in self.file_names.items():
+            file_path = os.path.join(raw_dataset_dir, file_name)
+            print("file: %s" % key)
+            f = open(file_path)
+            # read the definition of idx1-ubyte and idx3-ubyte
+            f.seek(4)
+            num = f.read(4)
+            num = int().from_bytes(num, "big")
+            print("size of %s : %d" % (key, num))
+            if re.match(r".*data.*", key) is not None:
+                height = f.read(4)
+                height = int().from_bytes(height, "big")
+                width = f.read(4)
+                width = int().from_bytes(width, "big")
+
+                data = np.frombuffer(f.read(), np.uint8).reshape(num, height, width)
+
+                # PCA projecting with 75% energy removing
+                n_comp = int(height * width)
+                pca = PCA(n_components=n_comp)
+                projected = pca.fit_transform(data.reshape(num, height * width))
+                n_comp = ((np.cumsum(pca.explained_variance_ratio_) > 0.25) != 0).argmax()
+                rec = np.matmul(projected[:, :n_comp], pca.components_[:n_comp])
+
+                saved_path = os.path.join(raw_dataset_dir, "images")
+                if not os.path.exists(saved_path):
+                    os.makedirs(saved_path)
+                saved_name = key + ".npy"
+                np.save(os.path.join(saved_path, saved_name), rec)
+            else:
+                data = np.frombuffer(f.read(), np.uint8)
+
+                saved_path = os.path.join(raw_dataset_dir, "labels")
+                if not os.path.exists(saved_path):
+                    os.makedirs(saved_path)
```

Comment:
nit: since you've imported `makedirs` from `fs_utils.py`, should we use that everywhere instead of `os.makedirs`? That should also account for other filesystems like S3 etc if users want to use them / if we want to allow passing in non local filesystem paths in the future!

Reply:
Updated